File: Gene/TDD-By-Example/Money.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TDD_By_Example
{
    abstract class Money
    {
        protected int Amount;
        protected string Currency;
        abstract public Money Times(int multiplier);

        static public Money dollar(int amount)
        {
            return new Dollar(amount);
        }

        static public Franc franc(int amount)
        {
            return new Franc(amount);
        }

        public override bool Equals(object obj)
        {
            Money money = (Money)obj;
            return this.Amount == money.Amount
                && this.GetType().Name.Equals(money.GetType().Name);
        }

        public string GetCurrency()
        {
            return Currency;
        }
    }
}

File: titan/src/money.ts
import { Bank } from './bank';
import { Expression } from './expression';
import { Sum } from './sum';

export class Money implements Expression {
  constructor(public amount: number, protected currency: string) {}

  static franc(amount: number): Money {
    return new Money(amount, 'CHF');
  }

  static dollar(amount: number): Money {
    return new Money(amount, 'USD');
  }

  toString(): string {
    return `${this.amount} ${this.currency}`;
  }

  times(multiplier: number): Money {
    return new Money(this.amount * multiplier, this.currency);
  }

  getCurrency(): string {
    return this.currency;
  }

  equals(object: object): boolean {
    const money: Money = object as Money;
    return (
      this.amount === money.amount && this.getCurrency() === money.getCurrency()
    );
  }

  plus(addend: Expression): Expression {
    return new Sum(this, addend);
  }

  reduce(bank: Bank, to: string): Money {
    const rate: number = bank.rate(this.currency, to);
    return new Money(this.amount / rate, to);
  }
}
